### ISSUE TITLE
How does masking affect forward and back propagation through time in LSTMs?

### ISSUE THREAD

**Issue by jaweriaamjad** [keras-team/keras#18211]
I am using Tensorflow API to code variational autoencoder for time series data. My model is as follows:

```def get_model(max_ts, padding_value):
    # encoder
    inp_x = tfk.layers.Input(shape=(max_ts, 1))
    masked_x = tfk.layers.Masking(mask_value=padding_value)(inp_x) 
    x = tfk.layers.LSTM(64,return_sequences=True)(masked_x)
    x = tfk.layers.LSTM(64)(x)
    z = tfk.layers.Dense(32, activation="relu")(x)
    z_mean = tfk.layers.Dense(latent_dim)(z)
    z_log_sigma = tfk.layers.Dense(latent_dim)(z)
    encoder = tfk.Model(inp_x, [z_mean, z_log_sigma])
    
    # decoder 
    inp_z = tfk.layers.Input(shape=(latent_dim,))
    y = masked_repeatvector(max_ts, padding_value)([inp_z, inp_x])
    y = tfk.layers.LSTM(64, return_sequences=True)(y)
    out = tfk.layers.TimeDistributed(tfk.layers.Dense(1))(y)
    decoder = tfk.Model(inp_z, out)   
    
    # encoder & decoder 
    z_mean, z_log_sigma = encoder(inp_x)
    z = tfk.layers.Lambda(sampling)([z_mean, z_log_sigma])
    out = decoder(z)
    vae = tfk.Model(inp_x, out)
    vae.add_loss(vae_loss(inp_x, out, max_ts, z_log_sigma, z_mean))
    # vae.add_loss(masked_vae_loss(inp_x, out, max_ts, z_log_sigma, z_mean, padding_value))
    return vae, encoder, decoder
```

As you can see my data is variable length so naturally I am using padding at the input. however, the padding information is lost at the end of the decoder. So to ensure the decoder accounts for the padding values, I have two options 1) either include a masking layer after the repeatvector layer 2) include a masked loss. I have gone with the second solution. My loss function is as follow:

```
def masked_vae_loss(original, out, ts,  z_log_sigma, z_mean, padding_value):
    mask = tf.cast(tf.not_equal(original, padding_value), dtype=tf.float32)
    reconstruction = K.mean(K.square(tf.math.multiply(original - out,mask)))*ts
    kl = -0.5 * K.mean(1 + z_log_sigma - K.square(z_mean) - K.exp(z_log_sigma))
    return reconstruction + kl
``` 

My model works but I get outputs with last value repeated for the padded values. Same thing happens when i use the decoder to generate timeseries data from sampled latents. I am having trouble understanding the outputs. From my research, I know that in Tensorflow, LSTM has been implemented so that last hidden states is copied for the padded/masked values. Does it also happen when the loss is maksed and no explicit masking is provided?

**Comment by tilakrayal** [comment]
@jaweriaamjad,
Could you please provide the complete standalone code or colab gist to reproduce the issue, which also helps us to debug the issue in an effective way. Thank you!

**Comment by jaweriaamjad** [comment]
Hi! Please find the link to the colab gist:

https://gist.github.com/jaweriaamjad/f54c6aa65cf52dcef7ee5b408e13af4d

**Comment by tilakrayal** [comment]
@jaweriaamjad,
Apologies for the delay. I tried to execute the mentioned code on tf-nightly and found that outputs with the last value were not repeated for the padded values. Kindly find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/42f8c9cee700b923c5817484019b7a4f/lstmvae.ipynb). 

1. Masking is not that complicated if we understand how the loss is computed with masking. For instance let us assume we have a sequence with length 256. From this sequence we have a masking with only 4 elements that are with masking of 1 (others are with masking 0). I thought the loss is computed as the average between these 4 elements. Guess what - it is not! The average loss will be divided by 256 instead. For this reason sometimes the loss will be extremely small (0.0something) if we have only few 1 elements and long sequence.
Does it matter? I guess not, as what we need is the gradient of loss, rather than the loss itself.

2. When we use softmax as the last layer, the denominator would be the sum of exponential of all elements, regarding whether their masking is 1 or 0.

3. I thought the output of masking inputs is zeros all the time in LSTM. But it is not the case. Let us assume we have a masking:

0 0 0 1 1 0 0 0

4. With this case, the three first elements with masking zero has output of 0. However, the three last zeros have output that is as the same as the output of the last element with masking 1.

5. Meanwhile, Keras is very convenient in the sense that the loss it computes will be based on only elements with masking of 1. I found this is a big plus of using Keras, something a bit too good too be true as I guess implementing this is not that easy.


Also each Keras layer declares if it supports masking. Each layer is also responsible for using the mask in a sensible way (which I believe is the primary source of confusion: that the masking functionality is implemented across a bunch of different classes). For RNN layers in particular, they rely on the fact that the underlying K.rnn operation has mask support so if you're looking for where precisely the logic is, you'll note that [the RNN layers simply pass the mask argument into the backend](https://github.com/fchollet/keras/blob/c0d185b467a569cc066372dd37837fbdb3641e6a/keras/layers/recurrent.py#L295), where [the magic happens](https://github.com/fchollet/keras/blob/c0d185b467a569cc066372dd37837fbdb3641e6a/keras/backend/tensorflow_backend.py#L2202-L2243).
